bpf/init.sh

```diff
@@ -296,16 +296,49 @@ function bpf_clear_cgroups()
 	HOOK=$2
 	NAME=$3
 
+	# Since Linux commit 1ba5ad36e00f ("bpftool: Use libbpf_bpf_attach_type_str"),
+	# bpftool uses the libbpf_bpf_attach_type_str() format in 'bpftool cgroup
+	# show' output. Perform a naive translation to ensure compatibility with prior
+	# bpftool versions and to avoid updating the hook name at all call sites. The
+	# transformed string can be used in a suffix match against the new format.
+
+	# Examples: (old -> transformed, new)
+	# connect4 -> 4_connect, cgroup_inet4_connect
+	# post_bind6 -> 6_post_bind, cgroup_inet6_post_bind
+	# sendmsg4 -> 4_sendmsg, cgroup_udp4_sendmsg
+
+	# There is no inet4/6 variant of e.g. sendmsg, so it's safe to assume the
+	# intended udp4/6 hook is selected.
+	newhook=$(echo "$HOOK" | sed -E 's/([a-z_]+)([0-9])/\2_\1/')
+
 	set +e
 	bpftool link detach pinned "$BPFFS_ROOT/cilium/socketlb/links/cgroup/$NAME" || true
 	rm -f "$BPFFS_ROOT/cilium/socketlb/links/cgroup/$NAME"
-
-	ID=$(bpftool cgroup show $CGRP | grep -w $HOOK | awk '{print $1}')
 	set -e
-	
-	if [ -n "$ID" ]; then
-		bpftool cgroup detach $CGRP $HOOK id $ID
-	fi
+
+	# Get all programs attached to the given cgroup and store their ids in a
+	# space-separated string. Run jq in raw output mode (-r) to prevent the
+	# resulting string from being written to stdout with double quotes. Perform a
+	# full match on the 'legacy' hook name appearing in older versions of bpftool,
+	# but perform a suffix match using the 'new' hook name.
+	ids=$(bpftool cgroup show "$CGRP" -j |
+		jq -r --arg legacy "$HOOK" --arg new "$newhook" '. |
+			map(. |
+				select(
+					.attach_type == $legacy or
+					(.attach_type | endswith($new))
+				)
+				| .id)
+			| join(" ")')
+
+	# Cilium versions 1.14 and later use ebpf-go to attach cgroup programs, which
+	# potentially attaches programs using the 'multi' flag if the kernel is recent
+	# enough to support the flag, but too old to support bpf_link. Detach all
+	# programs at the given hook since we can't reliably determine which ones
+	# Cilium owns.
+	for id in $ids; do
+		bpftool cgroup detach "$CGRP" "$HOOK" id "$id"
+	done
 }
```

Comment:
When I read this, I see a bug report coming: "all my BPF programs get randomly unloaded and I can't tell why but it feels like Cilium is doing it". Can we mitigate this somehow? Even if we need a newer kernel in order to detect Cilium programs.

Reply:
Hmm, based on context from the second commit I gather that there's no way for Cilium to cooperate with other programs on these hooks at the moment anyway, so if we don't delete those programs then Cilium will fail to load the programs.

I suspect we could be a bit more nuanced in our response here if this were in Go code rather than in the bash script, but I do wonder if we should be trying to detect the presence of other programs & trying to raise the alarm in such cases to inform the user that such a combination is not really supported.

But, in the mean time, I'm not immediately coming up with any better solutions than what you're proposing here. :+1:

Reply:
> Hmm, based on context from the second commit I gather that there's no way for Cilium to cooperate with other programs on these hooks at the moment anyway, so if we don't delete those programs then Cilium will fail to load the programs.

Yep, indeed. In all versions of Cilium to date, the agent has always replaced whatever program was attached before, if any, regardless of provenance.

However, it would only succeed at this if the preexisting program was attached without flags. If it was attached wifh flags, the replacement semantics of the API change completely, and a different approach is needed. That's what these patches try to ensure going forward, because in 1.14, the picture is slightly changing, more below.

> I suspect we could be a bit more nuanced in our response here if this were in Go code rather than in the bash script. I wonder if we should be trying to detect the presence of other programs & trying to raise the alarm in such cases to inform the user that such a combination is not really supported.

At a high level, there's no real difference in behaviour wrt. detaching existing progs. Cilium has always assumed full control over cgroup hooks, and that's not changing with these patches.

However, in the Go code proposed in the linked PR for 1.14, the happy path uses bpf_link, which sidesteps this whole mess. It attaches using the multi flag implicitly, so there can be other progs attached to the same hook and Cilium doesn't need to care. If a bpf_link was created and pinned, it won't try to detach anything else, and if a pinned link was found at a known path in bpffs, it'll simply update that link with the new program and call it a day. Much simpler.

The bad news is there are many corner cases (like upgrades) that can cause the initial bpf_link attachment to fail and require similar intervention to detach existing programs that were attached using the old API. The proposed Go code includes the same procedure as was written here in shell to make sure this all works without manual intervention, like rebooting the node.

Fortunately, the downgrade path is simple, since it cannot be done seamlessly. For the upgrade and agent restart path, we make sure the replacement is attached before detaching the old one if at all possible, to avoid missing events.

> But, in the mean time, I'm not immediately coming up with any better solutions than what you're proposing here. :+1: 

Thanks, it's taken us multiple months of iteration to get to this point. Making this sound across different kernel and Cilium versions was not straightforward. 😕 We'll have similar struggles migrating tc attachments from netlink to bpf_link, but that's still a few years out at this point.

Reply:
FWIW, this is only detaching BPF cgroup programs attached at the `/run/cilium/cgroupv2` path (and not the default `/sys/fs/cgroup` that typically has systemd attached programs). Maybe it's still worth printing what programs are being detached by the script for debugability reasons?

Reply:
> Maybe it's still worth printing what programs are being detached by the script for debugability reasons?

This is already the case due to `set -x`, no need to add more output. :slightly_smiling_face: